runtime/src/accounts_db.rs

```diff
@@ -2209,67 +2251,186 @@ impl AccountsDb {
+                // fetched from storage. This means a race occurred between this function and clean
+                // accounts/purge_slots
+                datapoint_warn!(
+                    "accounts_db-do_load_warn",
+                    (
+                        "warn",
+                        format!("do_load() failed to get key: {} from storage, latest attempt was for slot: {}, storage_entry: {}",
+                        pubkey,
+                        slot,
+                        store_id),
+                        String
+                    ),
+                );
+                return None;
+            }
+            match account_accessor {
+                LoadedAccountAccessor::Stored(None) => {
+                    if !is_root_fixed {
+                        // When running replay on the validator, or banking stage on the leader,
+                        // it should be very rare that the storage entry doesn't exist if the
+                        // entry in the accounts index is the latest version of this account.
+                        //
+                        // There are only a few places where the storage entry may not exist
+                        // after reading the index:
+                        // 1) Shrink has removed the old storage entry and rewritten to
+                        // a newer storage entry
+                        // 2) The `pubkey` asked for in this function is a zero-lamport account,
+                        // and the storage entry holding this account qualified for zero-lamport clean.
+                        //
+                        // In both these cases, it should be safe to retry and recheck the accounts
+                        // index. Also note that in both cases, if we do find the storage entry,
+                        // we can guarantee that the storage entry is safe to read from because
+                        // we grabbed a reference to the storage entry while it was still in the
+                        // storage map. This means even if the storage entry is removed from the storage
+                        // map after we grabbed the storage entry, the recycler should not reset the
+                        // storage entry until we drop the reference to the storage entry.
+
+                        // RPC get_account() may have fetched an old root from the index that was
+                        // either:
+                        // 1) Cleaned up by clean_accounts(), so the accounts index has been updated
+                        // and the storage entries have been removed.
+                        // 2) Dropped by purge_slots() because the slot was on a minor fork, which
+                        // removes the slots' storage entries but doesn't purge from the accounts index
+                        // (account index cleanup is left to clean for stored slots).
+                        num_acceptable_failed_iterations += 1;
+                    }
```

Comment:
if `is_root_fixed` is `true`, and `LoadedAccountAccessor::Stored(None)`, this could possibly lead into infinite loop. Anyway, this condition should never occur? how about like this?

```
    ...
} else {
    panic!("this shouldn't happen because of foo bar...");
}

```

Reply:
I think it can happen. Even during replay, you could have continuous shrink + clean of an existing storage entry, see comments here: https://github.com/solana-labs/solana/commit/e7ae7620f23cf849b16efa621f5498f0ef7813c1#diff-1090394420d51617f3233275c2b65ed706b35b53b115fe65f82c682af8134a6fR2289-R2296

Reply:
This is part of the reason I went with the `is_root_fixed` name, b/c if the root is fixed, there should be a bounded number of pending cleans/shrinks